src/software/amazon/ionschema/internal/util/RangeBigDecimal.kt

```diff
@@ -0,0 +1,120 @@
+package software.amazon.ionschema.internal.util
+
+import software.amazon.ion.*
+import software.amazon.ionschema.InvalidSchemaException
+import java.math.BigDecimal
+
+/**
+ * Implementation of Range<BigDecimal>.  As of this writing, all other Range<T>
+ * implementations directly or indirectly delegate to an instance of this class.
+ */
+internal class RangeBigDecimal(private val ion: IonList) : Range<BigDecimal> {
+    companion object {
+        private fun toBigDecimal(ion: IonValue) =
+                if (ion.isNullValue) {
+                    throw InvalidSchemaException("Unable to convert $ion to BigDecimal")
+                } else {
+                    when (ion) {
+                        is IonDecimal -> ion.bigDecimalValue()
+                        is IonFloat -> ion.bigDecimalValue()
+                        is IonInt -> BigDecimal(ion.bigIntegerValue())
+                        else ->
+                            throw InvalidSchemaException(
+                                    "Expected range lower/upper to be a decimal, float, or int (was $ion)")
+                    }
+                }
+    }
+
+    internal val lower: Boundary
+    internal val upper: Boundary
+    init {
+        if (!ion.hasTypeAnnotation("range")) {
```

Comment:
I'm not sure why none of the other `Range<T>` implementations do not have a similar check.  You can probably add this in the `init` block of a common base class which also defines a `protected val ion: IonList`) property.